src/modules/kvs/cache.c

```diff
@@ -228,44 +223,58 @@ int cache_entry_set_raw (struct cache_entry *hp, void *data, int len)
                 }
             }
         }
-        hp->type = CACHE_DATA_TYPE_RAW;
         return 0;
     }
     return -1;
 }
 
 json_t *cache_entry_get_json (struct cache_entry *hp)
 {
-    if (!hp || !hp->entry_valid || hp->type != CACHE_DATA_TYPE_JSON)
+    if (!hp || !hp->entry_valid || !hp->data)
         return NULL;
-    /* should be non-NULL for json */
-    assert (hp->o);
+    if (!hp->o) {
+        if (!(hp->o = json_loads (hp->data, JSON_DECODE_ANY, NULL)))
+            return NULL;
+    }
     return hp->o;
 }
 
 int cache_entry_set_json (struct cache_entry *hp, json_t *o)
 {
-    if (hp
-        && o
-        && (hp->type == CACHE_DATA_TYPE_NONE
-            || hp->type == CACHE_DATA_TYPE_JSON)) {
+    if (hp && o) {
         if (hp->entry_valid) {
-            assert (hp->o);
-            json_decref (o); /* no-op, 'o' is assumed identical to hp->o */
+            if (hp->data)
+                /* no-op, 'o' is assumed identical to current data */
+                json_decref (o);
+            else if (!hp->data) {
+                /* attempt to change already valid cache entry,
+                 * cannot, must call cache_entry_clear_data() */
```

Comment:
`cache_entry_clear_data()` doesn't appear to have any users

Reply:
Yeah created this function as a "just in case", knowing there were no real use cases at the time.  Perhaps would be wise to remove.  If anything, users can ```cache_remove_entry``` then recreate if they are in dire straits to put new data in the cache entry.